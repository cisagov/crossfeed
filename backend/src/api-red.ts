// Main entrypoint for red.
// Essentially, this entrypoint mimics API Gateway by passing down all requests
// to the crossfeed prod api lambda function.
import * as express from 'express';
import * as bodyParser from 'body-parser';
import * as cors from 'cors';
import { Lambda } from 'aws-sdk';

// Name of the lambda function name.
const API_LAMBDA_FUNCTION_NAME = 'crossfeed-prod-api';
const AWS_REGION = 'us-east-1';

interface LambdaResponse {
  isBase64Encoded: boolean;
  statusCode: number;
  headers: { [x: string]: string };
  multiValueHeaders: { [x: string]: string[] };
  body: string;
}

const app = express();
app.use(cors());
app.use(bodyParser.json());

app.use(async (req, res) => {
  // See https://docs.aws.amazon.com/apigateway/latest/developerguide/set-up-lambda-proxy-integrations.html#api-gateway-simple-proxy-for-lambda-input-format
  // for input / output formats for API gateway lambda integration.
  let body = JSON.stringify(req.body);
  if (body === '{}') {
    body = '';
  }
  const input = {
    resource: '/{any+}',
    path: req.path,
    httpMethod: req.method,
    headers: req.headers,
    multiValueHeaders: {},
    queryStringParameters: req.query,
    multiValueQueryStringParameters: {},
    pathParameters: {
      any: req.path.substr(1) // Remove leading slash from path
    },
    stageVariables: null,
    requestContext: {},
    body: body,
    isBase64Encoded: false
  };
  const lambdaClient = new Lambda({
    region: AWS_REGION
  });
  const response = await lambdaClient
    .invoke({
      FunctionName: API_LAMBDA_FUNCTION_NAME,
      InvocationType: 'RequestResponse',
      Payload: JSON.stringify(input)
    })
    .promise();
  const lambdaResponse: LambdaResponse = JSON.parse(
    response.Payload!.toString()
  );
  console.log(input.httpMethod, input.path, lambdaResponse.statusCode);
  if (lambdaResponse.statusCode >= 400) {
    console.error(
      'Request failed with status code',
      lambdaResponse.statusCode,
      'input',
      input,
      'response',
      lambdaResponse
    );
  }
  res.status(lambdaResponse.statusCode);
  for (const k in lambdaResponse.headers) {
    res.header(k, lambdaResponse.headers[k]);
  }
  res.send(lambdaResponse.body);
});

const port = 3000;
app.listen(port, () => {
  console.log('App listening on port ' + port);
});
