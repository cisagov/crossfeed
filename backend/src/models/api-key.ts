import {
  Entity,
  Column,
  PrimaryGeneratedColumn,
  UpdateDateColumn,
  CreateDateColumn,
  BaseEntity,
  ManyToOne
} from 'typeorm';
import { User } from './';

@Entity()
export class ApiKey extends BaseEntity {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @CreateDateColumn()
  createdAt: Date;

  @UpdateDateColumn()
  updatedAt: Date;

  @ManyToOne((type) => User, (user) => user.apiKeys, {
    onDelete: 'CASCADE',
    onUpdate: 'CASCADE'
  })
  user: User;

  @Column({
    type: 'timestamp',
    nullable: true
  })
  lastUsed: Date | null;

  @Column({
    // length based on hex formatted sha256
    length: 64,
    type: 'varchar'
  })
  hashedKey: string;

  @Column({
    length: 4,
    type: 'varchar'
  })
  lastFour: string;
}
