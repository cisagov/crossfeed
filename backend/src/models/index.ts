export * from './domain';
export * from './service';
export * from './connection';
export * from './vulnerability';
export * from './scan';
export * from './organization';
export * from './user';
export * from './role';
export * from './scan-task';
