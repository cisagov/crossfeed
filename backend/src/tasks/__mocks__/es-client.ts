export const syncDomainsIndex = jest.fn(() => ({}));

export const updateDomains = jest.fn(() => ({}));

export const searchDomains = jest.fn(() => ({}));

export default jest.fn(() => ({
  syncDomainsIndex,
  updateDomains,
  searchDomains
}));
