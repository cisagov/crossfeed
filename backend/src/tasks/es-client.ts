import { Client } from '@elastic/elasticsearch';
import { Domain, Webpage } from '../models';

export const DOMAINS_INDEX = 'domains-5';

interface DomainRecord extends Domain {
  suggest: { input: string | string[]; weight: number }[];
  parent_join: 'domain';
}

export interface WebpageRecord {
  webpage_id: string;
  webpage_createdAt: Date | null;
  webpage_updatedAt: Date | null;
  webpage_syncedAt: Date | null;
  webpage_lastSeen: Date | null;
  webpage_url: string;
  webpage_status: string | number;
  webpage_domainId: string;
  webpage_discoveredById: string;
  webpage_responseSize: number | null;
  webpage_headers: { name: string; value: string }[];

  // Added before elasticsearch insertion (not present in the database):
  suggest?: { input: string | string[]; weight: number }[];
  parent_join?: {
    name: 'webpage';
    parent: string;
  };
  webpage_body?: string;
}

/**
 * Elasticsearch client.
 */
class ESClient {
  client: Client;

  constructor(isLocal?: boolean) {
    this.client = new Client({ node: process.env.ELASTICSEARCH_ENDPOINT! });
  }

  /**
   * Creates the domains index, if it doesn't already exist.
   */
  async syncDomainsIndex() {
    const mapping = {
      services: {
        type: 'nested'
      },
      vulnerabilities: {
        type: 'nested'
      },
      webpage_body: {
        type: 'text',
        term_vector: 'yes'
      },
      parent_join: {
        type: 'join',
        relations: {
          domain: ['webpage']
        }
      }
    };
    try {
      await this.client.indices.get({
        index: DOMAINS_INDEX
      });
      await this.client.indices.putMapping({
        index: DOMAINS_INDEX,
        body: { properties: mapping }
      });
      console.log(`Index ${DOMAINS_INDEX} updated.`);
    } catch (e) {
      if (e.meta?.body?.error.type !== 'index_not_found_exception') {
        console.error(e.meta?.body);
        throw e;
      }
      await this.client.indices.create({
        index: DOMAINS_INDEX,
        body: {
          mappings: {
            properties: {
              ...mapping,
              suggest: {
                type: 'completion'
              }
            },
            dynamic: true
          },
          settings: {
            number_of_shards: 2
          }
        }
      });
      console.log(`Created index ${DOMAINS_INDEX}.`);
    }
    await this.client.indices.putSettings({
      index: DOMAINS_INDEX,
      body: {
        settings: { refresh_interval: '1800s' }
      }
    });
    console.log(`Updated settings for index ${DOMAINS_INDEX}.`);
  }

  excludeFields = (domain: Domain) => {
    const copy: any = domain;
    delete copy.censysCertificatesResults;
    for (const service in copy.services) {
      delete copy.services[service].censysIpv4Results;
      delete copy.services[service].wappalyzerResults;
      delete copy.services[service].intrigueIdentResults;
    }
    return copy;
  };

  /**
   * Updates the given domains, upserting as necessary.
   * @param domains Domains to insert.
   */
  async updateDomains(domains: Domain[]) {
    const domainRecords = domains.map((e) => ({
      ...this.excludeFields(e),
      suggest: [{ input: e.name, weight: 1 }],
      parent_join: 'domain'
    })) as DomainRecord[];
    const b = this.client.helpers.bulk<DomainRecord>({
      datasource: domainRecords,
      onDocument(domain) {
        return [
          {
            update: {
              _index: DOMAINS_INDEX,
              _id: domain.id
            }
          },
          { doc_as_upsert: true }
        ];
      },
      onDrop(doc) {
        console.error(doc.error, doc.document);
        b.abort();
      }
    });
    const result = await b;
    if (result.aborted) {
      console.error(result);
      throw new Error('Bulk operation aborted');
    }
    return result;
  }

  /**
   * Updates the given webpages, upserting as necessary.
   * @param webpages Webpages to insert.
   */
  async updateWebpages(webpages: WebpageRecord[]) {
    const webpageRecords = webpages.map((e) => ({
      ...e,
      suggest: [{ input: e.webpage_url, weight: 1 }],
      parent_join: {
        name: 'webpage',
        parent: e.webpage_domainId
      }
    })) as WebpageRecord[];
    const b = this.client.helpers.bulk<WebpageRecord>({
      datasource: webpageRecords,
      onDocument(webpage) {
        return [
          {
            update: {
              _index: DOMAINS_INDEX,
              _id: 'webpage_' + webpage.webpage_id,
              routing: webpage.webpage_domainId
            }
          },
          { doc_as_upsert: true }
        ];
      },
      onDrop(doc) {
        console.error(doc.error, doc.document);
        b.abort();
      }
    });
    const result = await b;
    if (result.aborted) {
      console.error(result);
      throw new Error('Bulk operation aborted');
    }
    return result;
  }

  /**
   * Searches for domains.
   * @param body Elasticsearch query body.
   */
  async searchDomains(body: any) {
    return this.client.search({
      index: DOMAINS_INDEX,
      body
    });
  }

  /**
   * Deletes everything in Elasticsearch
   */
  async deleteAll() {
    await this.client.indices.delete({
      index: '*'
    });
  }
}

export default ESClient;
