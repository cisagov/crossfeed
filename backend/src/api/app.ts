import * as express from 'express';
import * as bodyParser from 'body-parser';
import * as cookieParser from 'cookie-parser';
import * as cookie from 'cookie';
import * as cors from 'cors';
import * as helmet from 'helmet';
import { handler as healthcheck } from './healthcheck';
import { handler as scheduler } from '../tasks/scheduler';
import * as auth from './auth';
import * as domains from './domains';
import * as search from './search';
import * as vulnerabilities from './vulnerabilities';
import * as organizations from './organizations';
import * as scans from './scans';
import * as users from './users';
import * as scanTasks from './scan-tasks';
import * as stats from './stats';
import { listenForDockerEvents } from './docker-events';
import { createProxyMiddleware } from 'http-proxy-middleware';

if (
  (process.env.IS_OFFLINE || process.env.IS_LOCAL) &&
  typeof jest === 'undefined'
) {
  listenForDockerEvents();

  setInterval(() => scheduler({}, {} as any, () => null), 30000);
}

const handlerToExpress = (handler) => async (req, res, next) => {
  const { statusCode, body } = await handler(
    {
      pathParameters: req.params,
      requestContext: req.requestContext,
      body: JSON.stringify(req.body || '{}'),
      headers: req.headers,
      path: req.originalUrl
    },
    {}
  );
  try {
    const parsedBody = JSON.parse(body);
    res.status(statusCode).json(parsedBody);
  } catch (e) {
    // Not a JSON body
    res.setHeader('content-type', 'text/plain');
    res.status(statusCode).send(body);
  }
};

const app = express();

app.use(cors());
app.use(bodyParser.json());
app.use(helmet.hsts());
app.use(cookieParser());

app.get('/', handlerToExpress(healthcheck));
app.post('/auth/login', handlerToExpress(auth.login));
app.post('/auth/callback', handlerToExpress(auth.callback));

const checkUserLoggedIn = async (req, res, next) => {
  req.requestContext = {
    authorizer: await auth.authorize({
      authorizationToken: req.headers.authorization
    })
  };
  if (
    !req.requestContext.authorizer.id ||
    req.requestContext.authorizer.id === 'cisa:crossfeed:anonymous'
  ) {
    return res.status(401).send('Not logged in');
  }
  return next();
};

const checkUserSignedTerms = (req, res, next) => {
  // Bypass ToU for CISA emails
  const approvedEmailAddresses = ['@cisa.dhs.gov'];
  if (process.env.NODE_ENV === 'test')
    approvedEmailAddresses.push('@crossfeed.cisa.gov');
  for (const email of approvedEmailAddresses) {
    if (
      req.requestContext.authorizer.email &&
      req.requestContext.authorizer.email.endsWith(email)
    )
      return next();
  }
  if (
    !req.requestContext.authorizer.dateAcceptedTerms ||
    (req.requestContext.authorizer.acceptedTermsVersion &&
      req.requestContext.authorizer.acceptedTermsVersion !==
        getToUVersion(req.requestContext.authorizer))
  ) {
    return res.status(403).send('User must accept terms of use');
  }
  return next();
};

const getMaximumRole = (user) => {
  if (user?.userType === 'globalView') return 'user';
  return user && user.roles && user.roles.find((role) => role.role === 'admin')
    ? 'admin'
    : 'user';
};

const getToUVersion = (user) => {
  return `v${process.env.REACT_APP_TERMS_VERSION}-${getMaximumRole(user)}`;
};

// Rewrite the URL for some Matomo admin dashboard URLs, due to a bug in
// how Matomo handles relative URLs when hosted on a subpath.
app.get('/plugins/Morpheus/images/logo.svg', (req, res) =>
  res.redirect('/matomo/plugins/Morpheus/images/logo.svg?matomo')
);
app.get('/index.php', (req, res) => res.redirect('/matomo/index.php'));

const matomoProxy = createProxyMiddleware({
  target: process.env.MATOMO_URL,
  headers: { HTTP_X_FORWARDED_URI: '/matomo' },
  pathRewrite: function (path, req) {
    return path.replace(/^\/matomo/, '');
  },
  onProxyReq: function (proxyReq, req, res) {
    // Only pass the MATOMO_SESSID cookie to Matomo.
    const cookies = cookie.parse(proxyReq.getHeader('Cookie'));
    const newCookies = cookie.serialize(
      'MATOMO_SESSID',
      String(cookies['MATOMO_SESSID'])
    );
    proxyReq.setHeader('Cookie', newCookies);
  },
  onProxyRes: function (proxyRes, req, res) {
    // Remove transfer-encoding: chunked responses, because API Gateway doesn't
    // support chunked encoding.
    if (proxyRes.headers['transfer-encoding'] === 'chunked') {
      proxyRes.headers['transfer-encoding'] = '';
    }
  }
});

app.use(
  '/matomo',
  async (req, res, next) => {
    // Public paths -- see https://matomo.org/docs/security-how-to/
    const ALLOWED_PATHS = ['/matomo.php', '/matomo.js'];
    if (ALLOWED_PATHS.indexOf(req.path) > -1) {
      return next();
    }
    // API Gateway isn't able to proxy fonts properly -- so we're using a CDN instead.
    if (req.path === '/plugins/Morpheus/fonts/matomo.woff2') {
      return res.redirect(
        'https://cdn.jsdelivr.net/gh/matomo-org/matomo@3.14.1/plugins/Morpheus/fonts/matomo.woff2'
      );
    }
    if (req.path === '/plugins/Morpheus/fonts/matomo.woff') {
      return res.redirect(
        'https://cdn.jsdelivr.net/gh/matomo-org/matomo@3.14.1/plugins/Morpheus/fonts/matomo.woff'
      );
    }
    if (req.path === '/plugins/Morpheus/fonts/matomo.ttf') {
      return res.redirect(
        'https://cdn.jsdelivr.net/gh/matomo-org/matomo@3.14.1/plugins/Morpheus/fonts/matomo.ttf'
      );
    }
    // Only allow global admins to access all other paths.
    const user = (await auth.authorize({
      authorizationToken: req.cookies['crossfeed-token']
    })) as auth.UserToken;
    if (user.userType !== 'globalAdmin') {
      return res.status(401).send('Unauthorized');
    }
    return next();
  },
  matomoProxy
);

// Routes that require an authenticated user, without
// needing to sign the terms of service yet
const authenticatedNoTermsRoute = express.Router();
authenticatedNoTermsRoute.use(checkUserLoggedIn);
authenticatedNoTermsRoute.get('/users/me', handlerToExpress(users.me));
authenticatedNoTermsRoute.post(
  '/users/me/acceptTerms',
  handlerToExpress(users.acceptTerms)
);
authenticatedNoTermsRoute.put('/users/:userId', handlerToExpress(users.update));

app.use(authenticatedNoTermsRoute);

// Routes that require an authenticated user that has
// signed the terms of service
const authenticatedRoute = express.Router();
authenticatedRoute.use(checkUserLoggedIn);
authenticatedRoute.use(checkUserSignedTerms);

authenticatedRoute.post('/search', handlerToExpress(search.search));
authenticatedRoute.post('/domain/search', handlerToExpress(domains.list));
authenticatedRoute.post('/domain/export', handlerToExpress(domains.export_));
authenticatedRoute.get('/domain/:domainId', handlerToExpress(domains.get));
authenticatedRoute.post(
  '/vulnerabilities/search',
  handlerToExpress(vulnerabilities.list)
);
authenticatedRoute.post(
  '/vulnerabilities/export',
  handlerToExpress(vulnerabilities.export_)
);
authenticatedRoute.get(
  '/vulnerabilities/:vulnerabilityId',
  handlerToExpress(vulnerabilities.get)
);
authenticatedRoute.put(
  '/vulnerabilities/:vulnerabilityId',
  handlerToExpress(vulnerabilities.update)
);
authenticatedRoute.get('/scans', handlerToExpress(scans.list));
authenticatedRoute.get('/granularScans', handlerToExpress(scans.listGranular));
authenticatedRoute.post('/scans', handlerToExpress(scans.create));
authenticatedRoute.get('/scans/:scanId', handlerToExpress(scans.get));
authenticatedRoute.put('/scans/:scanId', handlerToExpress(scans.update));
authenticatedRoute.delete('/scans/:scanId', handlerToExpress(scans.del));
authenticatedRoute.post('/scans/:scanId/run', handlerToExpress(scans.runScan));
authenticatedRoute.post(
  '/scheduler/invoke',
  handlerToExpress(scans.invokeScheduler)
);
authenticatedRoute.post('/scan-tasks/search', handlerToExpress(scanTasks.list));
authenticatedRoute.post(
  '/scan-tasks/:scanTaskId/kill',
  handlerToExpress(scanTasks.kill)
);
authenticatedRoute.get(
  '/scan-tasks/:scanTaskId/logs',
  handlerToExpress(scanTasks.logs)
);

authenticatedRoute.get('/organizations', handlerToExpress(organizations.list));
authenticatedRoute.get(
  '/organizations/public',
  handlerToExpress(organizations.listPublicNames)
);
authenticatedRoute.get(
  '/organizations/:organizationId',
  handlerToExpress(organizations.get)
);
authenticatedRoute.post(
  '/organizations',
  handlerToExpress(organizations.create)
);
authenticatedRoute.put(
  '/organizations/:organizationId',
  handlerToExpress(organizations.update)
);
authenticatedRoute.delete(
  '/organizations/:organizationId',
  handlerToExpress(organizations.del)
);
authenticatedRoute.post(
  '/organizations/:organizationId/roles/:roleId/approve',
  handlerToExpress(organizations.approveRole)
);
authenticatedRoute.post(
  '/organizations/:organizationId/roles/:roleId/remove',
  handlerToExpress(organizations.removeRole)
);
authenticatedRoute.post(
  '/organizations/:organizationId/granularScans/:scanId/update',
  handlerToExpress(organizations.updateScan)
);
authenticatedRoute.post('/stats', handlerToExpress(stats.get));
authenticatedRoute.get('/users', handlerToExpress(users.list));
authenticatedRoute.post('/users', handlerToExpress(users.invite));
authenticatedRoute.delete('/users/:userId', handlerToExpress(users.del));

app.use(authenticatedRoute);

export default app;
