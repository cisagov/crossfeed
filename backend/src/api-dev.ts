// Main entrypoint for dev server.

import app from './api/app';

process.env.IS_OFFLINE = 'true';

const port = 3000;
app.listen(port, () => {
  console.log('App listening on port ' + port);
});
