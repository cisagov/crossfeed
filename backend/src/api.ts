// Main entrypoint for serverless API code.

import * as serverless from 'serverless-http';
import app from './api/app';

module.exports.handler = serverless(app, {
  binary: ['image/*', 'font/*']
});
