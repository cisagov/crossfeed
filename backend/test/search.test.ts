import * as request from 'supertest';
import app from '../src/api/app';
import { User, Domain, connectToDatabase, Organization } from '../src/models';
import { createUserToken } from './util';

import '../src/tasks/es-client';

jest.mock('../src/tasks/es-client');
const searchDomains = require('../src/tasks/es-client')
  .searchDomains as jest.Mock;
jest.mock('../src/api/search/buildRequest');
const buildRequest = require('../src/api/search/buildRequest')
  .buildRequest as jest.Mock;

describe('search', () => {
  let organization;
  beforeAll(async () => {
    await connectToDatabase();
    organization = await Organization.create({
      name: 'test-' + Math.random(),
      rootDomains: ['test-' + Math.random()],
      ipBlocks: [],
      isPassive: false
    }).save();
  });
  describe('search', () => {
    it('search by global admin should work', async () => {
      searchDomains.mockImplementation(() => ({
        body: [1, 2, 3]
      }));
      const response = await request(app)
        .post('/search')
        .set(
          'Authorization',
          createUserToken({
            userType: 'globalAdmin'
          })
        )
        .send({
          current: 1,
          resultsPerPage: 25,
          searchTerm: 'term',
          sortDirection: 'asc',
          sortField: 'name',
          filters: []
        })
        .expect(200);
      expect(buildRequest.mock.calls[0][1]).toEqual({
        matchAllOrganizations: true,
        organizationIds: []
      });
      expect(response.body).toEqual([1, 2, 3]);
    });
    it('search by regular user should work', async () => {
      const response = await request(app)
        .post('/search')
        .set(
          'Authorization',
          createUserToken({
            roles: [{ org: organization.id, role: 'user' }]
          })
        )
        .send({
          current: 1,
          resultsPerPage: 25,
          searchTerm: 'term',
          sortDirection: 'asc',
          sortField: 'name',
          filters: []
        })
        .expect(200);
      expect(buildRequest.mock.calls[0][1]).toEqual({
        matchAllOrganizations: false,
        organizationIds: [organization.id]
      });
    });
  });
});

describe('buildRequest', () => {
  const buildRequest = jest.requireActual('../src/api/search/buildRequest')
    .buildRequest;
  test('sample request by global admin', () => {
    const req = buildRequest(
      {
        current: 1,
        resultsPerPage: 25,
        searchTerm: 'term',
        sortDirection: 'asc',
        sortField: 'name',
        filters: []
      },
      { matchAllOrganizations: true, organizationIds: [] }
    );
    expect(req).toMatchSnapshot();
  });
  test('sample request by non-global admin', () => {
    const req = buildRequest(
      {
        current: 1,
        resultsPerPage: 25,
        searchTerm: 'term',
        sortDirection: 'asc',
        sortField: 'name',
        filters: []
      },
      { matchAllOrganizations: false, organizationIds: ['id1'] }
    );
    expect(req).toMatchSnapshot();
  });
});
