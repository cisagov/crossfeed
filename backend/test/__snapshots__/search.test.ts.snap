// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildRequest sample request by global admin 1`] = `
Object {
  "aggs": Object {
    "fromRootDomain": Object {
      "terms": Object {
        "field": "fromRootDomain.keyword",
      },
    },
    "name": Object {
      "terms": Object {
        "field": "name.keyword",
      },
    },
    "organization": Object {
      "terms": Object {
        "field": "organization.name.keyword",
      },
    },
    "services": Object {
      "aggs": Object {
        "name": Object {
          "terms": Object {
            "field": "services.service.keyword",
          },
        },
        "port": Object {
          "terms": Object {
            "field": "services.port",
          },
        },
        "products": Object {
          "aggs": Object {
            "cpe": Object {
              "terms": Object {
                "field": "services.products.cpe.keyword",
              },
            },
          },
          "nested": Object {
            "path": "products",
          },
        },
      },
      "nested": Object {
        "path": "services",
      },
    },
    "vulnerabilities": Object {
      "aggs": Object {
        "cve": Object {
          "terms": Object {
            "field": "vulnerabilities.cve.keyword",
          },
        },
        "severity": Object {
          "terms": Object {
            "field": "vulnerabilities.severity.keyword",
          },
        },
      },
      "nested": Object {
        "path": "vulnerabilities",
      },
    },
  },
  "highlight": Object {
    "fields": Object {
      "name": Object {},
    },
    "fragment_size": 200,
    "number_of_fragments": 1,
  },
  "query": Object {
    "bool": Object {
      "must": Array [
        Object {
          "match": Object {
            "parent_join": "domain",
          },
        },
        Object {
          "bool": Object {
            "should": Array [
              Object {
                "query_string": Object {
                  "analyze_wildcard": true,
                  "fields": Array [
                    "*",
                  ],
                  "query": "term",
                },
              },
              Object {
                "has_child": Object {
                  "inner_hits": Object {
                    "_source": Array [
                      "webpage_url",
                    ],
                    "highlight": Object {
                      "fields": Object {
                        "webpage_body": Object {},
                      },
                      "fragment_size": 50,
                      "number_of_fragments": 3,
                    },
                  },
                  "query": Object {
                    "query_string": Object {
                      "analyze_wildcard": true,
                      "fields": Array [
                        "*",
                      ],
                      "query": "term",
                    },
                  },
                  "type": "webpage",
                },
              },
            ],
          },
        },
      ],
    },
  },
  "size": 25,
  "sort": Array [
    Object {
      "name.keyword": "asc",
    },
  ],
}
`;

exports[`buildRequest sample request by non-global admin 1`] = `
Object {
  "aggs": Object {
    "fromRootDomain": Object {
      "terms": Object {
        "field": "fromRootDomain.keyword",
      },
    },
    "name": Object {
      "terms": Object {
        "field": "name.keyword",
      },
    },
    "organization": Object {
      "terms": Object {
        "field": "organization.name.keyword",
      },
    },
    "services": Object {
      "aggs": Object {
        "name": Object {
          "terms": Object {
            "field": "services.service.keyword",
          },
        },
        "port": Object {
          "terms": Object {
            "field": "services.port",
          },
        },
        "products": Object {
          "aggs": Object {
            "cpe": Object {
              "terms": Object {
                "field": "services.products.cpe.keyword",
              },
            },
          },
          "nested": Object {
            "path": "products",
          },
        },
      },
      "nested": Object {
        "path": "services",
      },
    },
    "vulnerabilities": Object {
      "aggs": Object {
        "cve": Object {
          "terms": Object {
            "field": "vulnerabilities.cve.keyword",
          },
        },
        "severity": Object {
          "terms": Object {
            "field": "vulnerabilities.severity.keyword",
          },
        },
      },
      "nested": Object {
        "path": "vulnerabilities",
      },
    },
  },
  "highlight": Object {
    "fields": Object {
      "name": Object {},
    },
    "fragment_size": 200,
    "number_of_fragments": 1,
  },
  "query": Object {
    "bool": Object {
      "must": Array [
        Object {
          "terms": Object {
            "organization.id.keyword": Array [
              "id1",
            ],
          },
        },
        Object {
          "bool": Object {
            "must": Array [
              Object {
                "match": Object {
                  "parent_join": "domain",
                },
              },
              Object {
                "bool": Object {
                  "should": Array [
                    Object {
                      "query_string": Object {
                        "analyze_wildcard": true,
                        "fields": Array [
                          "*",
                        ],
                        "query": "term",
                      },
                    },
                    Object {
                      "has_child": Object {
                        "inner_hits": Object {
                          "_source": Array [
                            "webpage_url",
                          ],
                          "highlight": Object {
                            "fields": Object {
                              "webpage_body": Object {},
                            },
                            "fragment_size": 50,
                            "number_of_fragments": 3,
                          },
                        },
                        "query": Object {
                          "query_string": Object {
                            "analyze_wildcard": true,
                            "fields": Array [
                              "*",
                            ],
                            "query": "term",
                          },
                        },
                        "type": "webpage",
                      },
                    },
                  ],
                },
              },
            ],
          },
        },
      ],
    },
  },
  "size": 25,
  "sort": Array [
    Object {
      "name.keyword": "asc",
    },
  ],
}
`;
